src/sonic-host-services/scripts/hostcfgd

```diff
@@ -49,13 +53,188 @@ def obfuscate(data):
+            (str): Target feature state
+        """
+
+        if state_configuration is None:
+            return None
+
+        template = jinja2.Template(state_configuration)
+        target_state = template.render(device_config)
+        if target_state not in ('enabled', 'disabled', 'always_enabled', 'always_disabled'):
+            raise ValueError('Invalid state rendered for feature {}: {}'.format(self.name, target_state))
+        return target_state
+
+
+class FeatureHandler(object):
+    """ Handles FEATURE table updates. """
+
+    SYSTEMD_SYSTEM_DIR = '/etc/systemd/system/'
+    SYSTEMD_SERVICE_CONF_D = os.path.join(SYSTEMD_SYSTEM_DIR, '{}.service.d/')
+
+    def __init__(self, config_db, device_config):
+        self._config_db = config_db
+        self._device_config = device_config
+        self._cached_config = {}
+        self.is_multi_npu = device_info.is_multi_npu()
+
+    def handle(self, feature_name, feature_cfg):
+        feature = Feature(feature_name, feature_cfg, self._device_config)
+        self._cached_config.setdefault(feature_name, Feature(feature_name, {}))
+
+        # Change auto-restart configuration first.
+        # If service reached failed state before this configuration applies (e.g. on boot)
+        # the next called self.update_feature_state will start it again. If it will fail
+        # again the auto restart will kick-in. Another order may leave it in failed state
+        # and not auto restart.
+        if self._cached_config[feature_name].auto_restart != feature.auto_restart:
+            self.update_feature_auto_restart(feature)
+            self._cached_config[feature_name].auto_restart = feature.auto_restart
+
+        # Enable/disable the container service if the feature state was changed from its previous state.
+        if self._cached_config[feature_name].state != feature.state:
+            self.update_feature_state(feature)
+            self._cached_config[feature_name].state = feature.state
+
+    def update_all_features_config(self):
+        feature_table = self._config_db.get_table('FEATURE')
+        for feature_name in feature_table.keys():
+            if not feature_name:
+                syslog.syslog(syslog.LOG_WARNING, "Feature is None")
+                continue
+
+            feature = Feature(feature_name, feature_table[feature_name], self._device_config)
+            self._cached_config.setdefault(feature_name, feature)
+
+            self.update_feature_auto_restart(feature)
+
+            self.update_feature_state(feature)
+            self.resync_feature_state(feature)
+
+    def update_feature_state(self, feature):
+        cached_feature = self._cached_config[feature.name]
+
+        if self.is_invariant_feature_state(cached_feature.state):
+            if cached_feature.state != feature.state:
+                syslog.syslog(syslog.LOG_INFO, "Feature {} service is {}".format(feature.name, feature.state))
+                self.resync_feature_state(cached_feature)
```

Comment:
Since the state of service/feature is updated to the new state `feature`, do we need pass `feature` to the function `resync_feature_state(feature)`? Can you check whether my understanding is correct please?

Reply:
Since the function `resync_feature_state(...)` is used to update the `state` of service/feature and has one line code, is it necessary to create a separate function to do this?

Reply:
> Since the function `resync_feature_state(...)` is used to update the `state` of service/feature and has one line code, is it necessary to create a separate function to do this?

For readability purpose

Reply:
@yozhao101, Please resolve or add your comments.
We need to get this PR in.

Reply:
> > Since the state of service/feature is updated to the new state `feature`, do we need pass `feature` to the function `resync_feature_state(feature)`? Can you check whether my understanding is correct or not please?
> 
> if current state is an invariant state and the new one is different then current one we resync with the current one. Please refer to original code: https://github.com/Azure/sonic-buildimage/pull/5744/files#diff-27d10088e7a11c3ae41f16e5272d96926179ae179c61d6fc32a25d0cfa40907eL497-L504

Please review the comment on the original code.